### ISSUE TITLE
How is the starting version calculated?

### ISSUE THREAD

**Issue by limonte** [jscutlery/semver#514]
If no releases of a package were made before, how is the starting version calculated? Is it taken from `package.json` or `v1.0.0` or something else?

I wasn't able to find anything about this in README.

**Comment by edbzn** [comment]
@limonte If no release is found, meaning that no tags exist, it will take the commit history since the first commit and calculate the version from it.

**Comment by edbzn** [comment]
Semver is tag-based, it never uses the version in package.json, it always parses the commit history and calculates the version from it.

**Comment by limonte** [comment]
Thank you for the answer. Let's assume there's one commit in the commit history: `feat: init my cool package`. Which version will be released?

**Comment by edbzn** [comment]
@limonte I would say 0.1.0.

**Comment by limonte** [comment]
> @limonte I would say 0.1.0.

That is what I'm experiencing in my project and that's unexpected. According to the semantic-release docs it should be `1.0.0`: https://semantic-release.gitbook.io/semantic-release/support/faq#can-i-set-the-initial-release-version-of-my-package-to-0.0.1

**Comment by limonte** [comment]
Another discussion suggests `1.0.0` https://github.com/semantic-release/semantic-release/issues/837

**Comment by edbzn** [comment]
Interesting I didn't know, I will investigate but it looks weird that we can't provide a feature in 0.x.x without bumping to 1.0.0.

0.x.x should only contain fixes?

**Comment by limonte** [comment]
This article is suggested by semver docs: https://blog.greenkeeper.io/introduction-to-semver-d272990c44f2

There's an explanation at the end about why it should be `1.0.0`

`0.x.x` could bring unexpected hard-to-understand troubles to developers. E.g. `^0.3.0` wouldn't be resolved to `0.4.0`, only to `0.3.x`.

**Comment by edbzn** [comment]
> 0.x.x could bring unexpected hard-to-understand troubles to developers. E.g. ^0.3.0 wouldn't be resolved to 0.4.0, only to 0.3.x.

You point out something I was not aware of, it is super confusing. On the other hand, starting the initial version from 1.0.0 would be an opinionated choice, I'm not against but we should document it well. @yjaaidi what do you think about that?

**Comment by limonte** [comment]
> On the other hand, starting the initial version from 1.0.0 would be an opinionated choice, I'm not against but we should document it well.

I support this behavior to be customizable, some developers might want to start with `0.x.x` 👍 But IMO the default behavior should be aligned with `semantic-release`, i.e. the starting version should be `1.0.0`.

**Comment by yjaaidi** [comment]
Thanks @limonte for raising this issue.
It is in fact a very interesting topic.

**TL;DR**: leading zeroes is not semantic versioning compliant but it is appealing 😊

It would be nice if we could figure this out and provide an opinionated approach. It would be nice to reduce the possible options surface.

What would be the right balance between semantic release spec and marketing? What provides the best DX?

Leading zeroes (i.e.. `0.x.y`) is not spec compliant but in terms of marketing it is more appealing than `1.0.0-rc.13`... the latest can sound less stable (while it's not).
Also even though it is not spec compliant `0.x.y` conventionally means `0.MAJOR|MINOR.PATCH` one can use `~0.x.y` and get the fixes while with prerelease, we would have to define a prerelease convention and complex `>=1.0.0-rc.1.0,<1.0.0-rc.2.0` matching or something like that.

So maybe, we could apply the following heuristic:
- if `releaseAs` is set as `prerelease` then we would start with `1.0.0-...`
- if `releaseAs` is set as `major` (probably through the CLI) then we would start at `1.0.0`
- otherwise, the default behavior would be to start with `0.1.0`

While we are at it, we should document the use of `releaseAs: prerelease` in `project.json` instead of on command line to encourage a descriptive approach instead of an imperative one.

What do you think?

**Comment by limonte** [comment]
@yjaaidi thank you for the detailed feedback! 

Your suggestion would mean this for our case:

1. create a commit with `--releaseAs=major` parameter
2. run CI workflow which will release `v1.0.0`
3. create another commit that will remove `--releaseAs=major` from the codebase and allow the next releases to be calculated based on the commit history

This doesn't look smooth to me or am I missing something?

**Comment by limonte** [comment]
Wouldn't it be handy to have the `--initialVersion` param (naming to be discussed) which will be `undefined` by default.

And then:

```js
if (params.initialVersion) {
  if (isValid(params.initialVersion)) {
    newVersion = params.initialVersion
  } else {
    throw new Error('Invalid initial version')
  }
} else {
  newVersion = calculateBasedOnCommitHistory()
}
```

WDYT?

**Comment by limonte** [comment]
> Actually, what I meant is that instead of `nx run workspace:version --initialVersion=1.0.0` you could simply call `nx run workspace:version --releaseAs=major`. What do you think?

I got you. I'm not running this command locally, but in CI because that's how it should be ideally done. So, in order to run it I need to have `nx run workspace:version --releaseAs=major` in the codebase, i.e. commit it to the releasing branch.

So the steps are:

1. commit `nx run workspace:version --releaseAs=major` to the releasing branch
2. run the releasing workflow in CI
3. make another commit which will remove `--releaseAs=major`

Is that correct or am I missing something?

---

On the other hand, with my suggestion with the `initialVersion` param, there wouldn't be need for step 3, because that param will only be used when there are no previous versions.

**Comment by yjaaidi** [comment]
Oh sorry! I forgot your first explanation when replying to your last message.
The only thing that I dislike about initial version option is how it pollutes the project json.
We can fix this by updating the project json and removing the option after the first release.

Now what should be the default?
Should we default to 1.0.0 to be compliant with the spec? @edbzn

**Comment by edbzn** [comment]
In fact, the semantic versioning spec allows `0.x.y` versions, here are the rules defined in the spec: 

- [Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable.](https://semver.org/spec/v0.1.0.html#spec-item-5)
- [Version 1.0.0 defines the public API. The way in which the version number is incremented is now dependent on this public API and how it changes.](https://semver.org/spec/v0.1.0.html#spec-item-6)

It's a choice made by the semantic-release team to start at `1.0.0`, but it's not spec-friendly. In my opinion, allowing `0.x.y` versions for initial development makes sense. I would just document the confusing usage of `^0.2.0`, and keep the calculation based on the commits history for the initial version, we can add documentation on how the initial version is calculated as well. 

If you want to start at `1.0.0` then I would recommend releasing manually with the option `--releaseAs=1.0.0`, I don't think we need to add a new option. Here again, we can document how to start at `1.0.0`.

**Comment by limonte** [comment]
First of all, I wouldn't expect recommendations for manual releasing from the authors of the plugin which is aimed to automate releasing and move it completely to CI.

Next, I don't think `--releaseAs=1.0.0` will work because the docs say:

> Level can be one of major, minor, patch, premajor, preminor, prepatch, or prerelease

So, I don't think `1.0.0` is a valid value for the `releaseAs` option.

---

What's the downside of having the additional param? Is it only polluting `project.json`?

**Comment by yjaaidi** [comment]
@limonte is there something preventing you from making the first release manually? Or is it just inconvenient?

**Comment by limonte** [comment]
> @limonte is there something preventing you from making the first release manually? Or is it just inconvenient?

In our team we want every team member to be able to make a release. For that, we have the dedicated CI workflow which can be triggered by anyone. 

We are releasing to GitHub Registry and in order for that `GITHUB_TOKEN` secret is used in that workflow.

For manual releasing we need to bother every team member with having that `GITHUB_TOKEN` locally which is inconvenient and certainly doesn't affect positively the security of that token.

**Comment by edbzn** [comment]
> Next, I don't think --releaseAs=1.0.0 will work because the docs say:

It is right the correct usage is `--releaseAs=major`. 

> I wouldn't expect recommendations for manual releasing from the authors of the plugin which is aimed to automate releasing and move it completely to CI.

1. You can automate the first release if you stick to the semver spec (ie starting at `0.x.y`).
2. We want to limit the public API surface as adding options comes with a maintenance cost. 
3. BUT if it turns out that it is useful for multiple users we can re-consider adding the `--initialVersion` option to avoid manual release, let's see if other ppl need it!

**Comment by yjaaidi** [comment]
This has to be verified but also, if the initial commit is marked as a breaking change, the first release should be 1.0.0 as the computed bump would be "major". This also a workaround.

The only thing that we don't cover yet is automating the initial version to a specific version like "3.1.0" because version management was handled in a different fashion before migrating to semver.

**Comment by why-this-code-works** [comment]
I'd consider my question fully answered, thank you both @edbzn and @yjaaidi for this discussion.

I'll close this issue and file a new one with the suggested feature request so people can easy find it and vote for it.

**Comment by limonte** [comment]
I'd consider my question fully answered, thank you both @edbzn and @yjaaidi for this discussion.

I'll close this issue and file a new one with the suggested feature request so people can easily find it and vote for it.

**Comment by piranna** [comment]
Honestly, I think it should start on `0.0.0`, at least until there's a first feature, in that case it makes sense to be `0.1.0`... First commits are usually just only to add scafolding files to prepare the environment, not actual development.
